### PR TITLE
Enable connection for assistant inner tool

### PR DESCRIPTION
**Pull request by chw-microsoft**
# Description

This PR is to support connection usage when you define the inner tools within assistant.

Changed involed: 
0) contract level: adding hinted attribute _tool_invoker as container for resolved structure
1) moved the assistant inner function resolve logic from AssistantToolInvoker to _tool_resolver.py and enable the connection type resolve
2) refined the assistant python script logic to handle better about state management
3) Create e2etests for assistant with connection run
4) Created connection for assistant aoai test

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/contracts/types.py [hunk 1]
```diff
@@ -29,13 +30,23 @@ class FilePath(str):
...
 class AssistantDefinition:
     """This class is used to define an assistant definition."""
 
     model: str
     instructions: str
-    tools: list
+    # The raw tool definition in json string
+    tools: List
+    # The resolved tool definition by name. This property is addon and will not be part of pickling.
+    assistant_tools: Dict[str, AssistantTool] = field(default_factory=dict)
```
Comment (thy09) [inline src/promptflow/promptflow/contracts/types.py:49]: I don't think it is a contract level concept, it is just about implementation, please try another way.

File: src/promptflow/promptflow/executor/_tool_resolver.py [hunk 2]
```diff
@@ -107,7 +110,79 @@ def _convert_to_assistant_definition(self, assistant_definition_path: str, input
...
+        func_name = resolved_tool.definition.function
+        definition = self._generate_tool_definition(func_name, resolved_tool.definition.description, predefined_inputs)
+        if resolved_tool.node.inputs:
+            inputs = {name: value.value for name, value in resolved_tool.node.inputs.items()}
+            func = partial(resolved_tool.callable, **inputs)
+        else:
+            func = resolved_tool.callable
+        return AssistantTool(name=func_name, openai_definition=definition, func=func)
+
+    def _create_node_for_assistant_tool(self, tool: dict):
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_tool_resolver.py:146]: Why should we rely on nodes concept for asssitant tool? I think it should be decoupled.

Comment (chw-microsoft) [reply src/promptflow/promptflow/executor/_tool_resolver.py:146]: Th reason is to reuse existing method for input resolve. Per offline discussion, we keep it as it is for now and refine this logic later

File: src/promptflow/promptflow/contracts/types.py (hunk 1 above)
Comment (chw-microsoft) [reply src/promptflow/promptflow/contracts/types.py]: Per offline discussion, we use implicit attribute to contain the resolved part

File: src/promptflow/promptflow/contracts/types.py [hunk 3]
```diff
@@ -50,7 +51,6 @@ def serialize(self):
             "tools": self.tools,
         }
 
-    def init_tool_invoker(self):
-        from promptflow.executor._assistant_tool_invoker import AssistantToolInvoker
-
-        return AssistantToolInvoker.init(self.tools)
+    def __post_init__(self):
+        # Implicitly introduce the 'tool_invoker' attribute here
+        self.tool_invoker = None  # reserved attribute for tool invoker injection
```
Comment (thy09) [inline src/promptflow/promptflow/contracts/types.py:56]: Use `_tool_invoker`

File: src/promptflow/promptflow/executor/_tool_resolver.py [hunk 4]
```diff
@@ -107,7 +111,78 @@ def _convert_to_assistant_definition(self, assistant_definition_path: str, input
...
+            else:
+                raise UnsupportedAssistantToolType(
+                    message_format="Unsupported assistant tool type: {tool_type}",
+                    tool_type=tool["type"],
+                    target=ErrorTarget.EXECUTOR,
+                )
+        assistant_definition.tool_invoker = AssistantToolInvoker(resolved_tools)
+
+    def _load_tool_as_function(self, tool: dict):
+        node, predefined_inputs = self._create_node_for_assistant_tool(tool)
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_tool_resolver.py:135]: We need to add a TODO here to decouple tool resolving and node resolving.

Comment (thy09) [reply src/promptflow/promptflow/executor/_tool_resolver.py:135]: We should also add more comments here to explain why we convert it to a node.

File: src/promptflow/promptflow/executor/_tool_resolver.py [hunk 5]
```diff
@@ -107,7 +111,78 @@ def _convert_to_assistant_definition(self, assistant_definition_path: str, input
...
+        )
+        return node, list(predefined_inputs.keys())
+
+    def _generate_tool_definition(self, func_name: str, description: str, predefined_inputs: list) -> dict:
+        try:
+            to_openai_type = {
+                "str": "string",
+                "int": "number",
+                "float": "number",
+                "bool": "boolean",
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_tool_resolver.py:165]: All these types are supported? I think we need to align tool input type.

File: src/promptflow/promptflow/executor/_tool_resolver.py [hunk 6]
```diff
@@ -107,7 +111,78 @@ def _convert_to_assistant_definition(self, assistant_definition_path: str, input
...
+            return {
+                "type": "function",
+                "function": {
+                    "name": func_name,
+                    "description": description,
+                    "parameters": {"type": "object", "properties": params, "required": list(params.keys())},
+                },
+            }
+        except Exception as e:
+            raise FailedToParseAssistantTool(func_name=func_name) from e
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_tool_resolver.py:185]: If we raise the exception here, all the tools without doc string cannot be used?

File: src/promptflow/promptflow/executor/_tool_resolver.py [hunk 7]
```diff
@@ -107,7 +111,78 @@ def _convert_to_assistant_definition(self, assistant_definition_path: str, input
...
+        node = Node(
+            name="assistant_node",
+            tool="assistant_tool",
+            inputs=predefined_inputs,
+            source=ToolSource.deserialize(tool["source"]) if "source" in tool else None,
+            type=ToolType.PYTHON if "tool_type" in tool and tool["tool_type"] == "python" else None,
+        )
+        return node, list(predefined_inputs.keys())
+
+    def _generate_tool_definition(self, func_name: str, description: str, predefined_inputs: list) -> dict:
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_tool_resolver.py:159]: Please add more UTs for different examples here.

File: src/promptflow/promptflow/executor/_tool_resolver.py (hunk 4 above)
Comment (chw-microsoft) [reply src/promptflow/promptflow/executor/_tool_resolver.py:135]: https://msdata.visualstudio.com/Vienna/_workitems/edit/2979976

File: src/promptflow/promptflow/executor/_tool_resolver.py (hunk 6 above)
Comment (chw-microsoft) [reply src/promptflow/promptflow/executor/_tool_resolver.py:188]: right, currently it is.  I think we used docstring to parse parameter +description. I created a working item to improve this:

File: src/promptflow/tests/executor/unittests/executor/test_assistant_tool_invoker.py [hunk 8]
```diff
@@ -1,104 +1,17 @@
...
-        assert "Unsupported assistant tool type" in exc_info.value.message
-
-    def _remove_predefined_inputs(self, value: any, predefined_inputs: list):
-        for input in predefined_inputs:
-            if input in value:
-                if isinstance(value, dict):
-                    value.pop(input)
-                elif isinstance(value, list):
-                    value.remove(input)
+from promptflow._sdk.entities import AzureOpenAIConnection
```
Comment (thy09) [inline src/promptflow/tests/executor/unittests/executor/test_assistant_tool_invoker.py:3]: `from promptflow.connections import `

File: src/promptflow/promptflow/executor/_tool_resolver.py (hunk 5 above)
Comment (chw-microsoft) [reply src/promptflow/promptflow/executor/_tool_resolver.py:168]: Per discussion with @thy09 , here is mapping from python type--> json data type.  Bit hard to invole promptflow ValueType. But we will refine the resolve logic for assistant soon. So leave this change for now.

File: src/promptflow/promptflow/executor/_tool_resolver.py (hunk 6 above)
Comment (chw-microsoft) [reply src/promptflow/promptflow/executor/_tool_resolver.py:188]: I think this exception would be deprecate or downgrade to be warning once we get independent of docstring for function parse

File: src/promptflow/promptflow/executor/_tool_resolver.py (hunk 7 above)
Comment (chw-microsoft) [reply src/promptflow/promptflow/executor/_tool_resolver.py:162]: sure, https://msdata.visualstudio.com/Vienna/_workitems/edit/2980105

File: src/promptflow/tests/executor/unittests/executor/test_assistant_tool_invoker.py (hunk 8 above)
Comment (chw-microsoft) [reply src/promptflow/tests/executor/unittests/executor/test_assistant_tool_invoker.py:3]: good catch

**Review by lumoslnt — APPROVED** [review]

**Pull request merged**